Satisfy testing on macOS

amotl opened the pull request:
## About
This is somehow related to https://github.com/crate/crash/issues/343, it makes the test suite work on macOS.

On top of using [conditional buildout configuration sections](http://www.buildout.org/en/latest/topics/variables-extending-and-substitutions.html#conditional-configuration-sections) for pulling down the appropriate CrateDB distribution tarballs for other operating systems than Linux, ~~it skips some tests currently not working on macOS~~. Windows support has not been verified yet, I am working on this on behalf of #388.

## Summary of the changes
- Remove doctests about `CrateLayer.from_uri` as it does not honor the platform when building the download URL. Instead, re-add it as a unit tests which at least runs the function by mocking `_download_and_extract` into a noop.
- Discriminate between testing against CrateDB nightly (run on `ubuntu-latest`) and multi-platform tests against a specific version of CrateDB (run on both `ubuntu-latest` and `macos-latest`).
- Improve stability by using `REFRESH TABLE` appropriately.
- Use most recent CrateDB 4.5.0.

amotl commented:
It looks like GHA is currently not able to access `cdn.crate.io` [1]. Downloading the latest release tarball croaks with `[Errno 104] Connection reset by peer`. Has anyone seen that before?

```
 Downloading https://cdn.crate.io/downloads/releases/nightly/crate-latest.tar.gz
While:
  Installing crate.
Error: Error downloading extends for URL https://cdn.crate.io/downloads/releases/nightly/crate-latest.tar.gz: [Errno 104] Connection reset by peer
Error: Process completed with exit code 1.
```

[1] https://github.com/crate/crate-python/pull/386/checks?check_run_id=1322270579

amotl commented:
> Downloading the latest CrateDB release tarball croaks with `[Errno 104] Connection reset by peer`.

Downloading the release tarballs started working again.

Now the `Test python-3.6 sqla: 1.1.15 CrateDB: nightly on ubuntu-latest` section croaks [1] with
```
Failed doctest test for https.txt
```
So, #384 will hopefully heal that?

[1] https://github.com/crate/crate-python/pull/386/checks?check_run_id=1322310763

amotl commented:
Now, all tests succeeded.

mfussenegger reviewed (commented):
Could we also add windows and macos to the build matrix in the github actions config?

amotl commented on base.cfg:
```diff
@@ -19,17 +19,34 @@ eggs = crate
+relative-paths=true
+working-directory = ${buildout:directory}
+recipe = zc.recipe.testrunner
+defaults = ['--auto-color', '--ignore_dir', 'testing', '--test', '!test_no_retry_on_read_timeout|test_client_threaded|test_client_keepalive|test_username|test_default_schema']
```
I didn't do that lightheartedly and I totally second your arguments in general. The main reason I've put it in there is that I didn't want to spend more time on that but tried to get a reasonable amount of tests working flawlessly on my machine.

Now, let's quickly have a look at the details behind the workaround.

1. When not using `--ignore_dir=testing`, the test suite would invoke `CrateLayer.from_uri()` from [`testing/doctests/layer.txt`](https://github.com/crate/crate-python/blob/0.26.0/src/crate/testing/doctests/layer.txt), which in turn is not prepared for running on macOS yet.

2. When running `./bin/test -vvvvv --test=test_no_retry_on_read_timeout`, I received
```
 test_no_retry_on_read_timeout (crate.client.test_http.RetryOnTimeoutServerTest)Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
  File "/Users/amo/dev/crate/sources/crate-python/src/crate/client/test_http.py", line 509, in <module>
    class TestingHTTPServer(HTTPServer):
  File "/Users/amo/dev/crate/sources/crate-python/src/crate/client/test_http.py", line 513, in TestingHTTPServer
    manager = multiprocessing.Manager()
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/managers.py", line 579, in start
    self._process.start()
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```
   I haven't been able to make any further sense of that.

amotl replied:
> I'm not keen on building workarounds for problems that should be solved.

#405 solved the root cause of this problem I was observing here, so I amended the patch here accordingly.

amotl commented:
> Could we also add windows and macos to the build matrix in the github actions config?

That would be nice, right?

amotl commented:
> Could we also add windows and macos to the build matrix in the github actions config?

I just added `macos-latest`. For Windows, I am working on #388.

Because I have not been able to adjust the test matrix appropriately to run CrateDB nightly on `ubuntu-latest`, but a specific version like e.g. CrateDB 4.5.0 on `macos-latest`, I split the GHA recipe into two halves: One for checking pull requests on all operating systems and another one for checking against CrateDB nightly on regular intervals as configured through the cronjob-like scheduling configuration. That latter exclusively happens on `ubuntu-latest`, as nightly builds are not available for other operating systems than Linux.
